RHDEVDOCS-5245: Performance Tuning for Openshift Pipelines (#60648)
openshift/openshift-docs: modules/op-performance-tuning-using-tektonconfig-cr.adoc

// This module is included in the following assembly:
//
// *openshift_pipelines/customizing-configurations-in-the-tektonconfig-cr.adoc

:_content-type: CONCEPT
[id="op-performance-tuning-using-tektonconfig-cr_{context}"]
= Performance tuning using TektonConfig CR

> You can modify the fields under the `.spec.performance` parameter in the `TektonConfig` custom resource (CR) to change high availability (HA) support and performance configurations for the {pipelines-shortname} controller.

Reviewer: `.spec.pipeline.performance` parameter in the `TektonConfig`

.Example: `TektonConfig` performance fields
[source,yaml]
----
apiVersion: operator.tekton.dev/v1alpha1
kind: TektonConfig
metadata:
  name: config
spec:
  performance:
    disable-ha: false
    buckets: 1
    threads-per-controller: 2
    kube-api-qps: 5.0

>     kube-api-burst: 10

Reviewer: ```yaml
spec:
  pipeline:
    performance:
      disable-ha: false
      buckets: 1
      threads-per-controller: 2
      kube-api-qps: 5.0
      kube-api-burst: 10
```

----

> The fields are optional. If you set them, the {pipelines-title} Operator includes most of the fields as arguments in the `tekton-pipelines-controller` deployment under the `tekton-pipelines-controller` container. The {pipelines-shortname} Operator also updates the `buckets` field in the `config-leader-election` configuration map under the `tekton-pipelines` namespace.

Reviewer: namespace in product will be `openshift-pipelines`
`tekton-pipelines` should be as `openshift-pipelines`

> If you do not specify the fields, the {pipelines-shortname} Operator applies the default values for the {pipelines-shortname} controller.

Reviewer: if you do not specify the values, operator will not update those fields. default values from pipelines-controller will be used.

[NOTE]
====
If you modify or remove any of the performance fields, the {pipelines-shortname} Operator updates the `tekton-pipelines-controller` deployment and the `config-leader-election` configuration map (if the `buckets` field changed), and recreates `tekton-pipelines-controller` pods.
====

> The following are the modifiable fields:

Author: I put it in the table, but for the table to not look squeezed, I had to move the notes outside of the table.

> The following are the modifiable fields:

Reviewer 2: All good on this one!

> * `disable-ha`: Enable or disable the HA support. The default value for the pipeline controller is `false`.

Reviewer 2: Why "for the pipeline controller"? And if the value of false means that HA is actually enabled by default, could you maybe explain this here?

Author: @mramendi I wrote "for the pipeline controller" to specify that the fields are optional. The default values therefore indicate what is the default configuration for the pipeline controller. Do you think the specification here is unnecessary?

* `buckets`: This field indicates the number of buckets used to partition key space of each reconciler.
+
The given number of replicas uses the specified number of buckets. The owner of a bucket handles the reconciling of the keys partitioned into that bucket. The maximum value is `10` and the default value in the pipeline controller is `1`.
+
[NOTE]
====
The {pipelines-shortname} Operator does not handle the pipeline controller deployment of `replicas`. For example, to change the number of replicas to 3, run:

[source,terminal]
----

> $ oc --namespace tekton-pipelines scale deployment tekton-pipelines-controller --replicas=3

Reviewer: namepsace => `openshift-pipelines`

// This module is included in the following assembly:
//
// *openshift_pipelines/customizing-configurations-in-the-tektonconfig-cr.adoc

:_content-type: CONCEPT
[id="op-performance-tuning-using-tektonconfig-cr_{context}"]
= Performance tuning using TektonConfig CR

You can modify the fields under the `.spec.pipeline.performance` parameter in the `TektonConfig` custom resource (CR) to change high availability (HA) support and performance configuration for the {pipelines-shortname} controller.

> .Example: `TektonConfig` performance fields

Reviewer 3: ```suggestion
.Example: TektonConfig performance fields
```
Headings can't have backticks or other markup. See https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#assemblymodule-titles-and-section-headings

      kube-api-qps: 5.0
      kube-api-burst: 10
----

The fields are optional. If you set them, the {pipelines-title} Operator includes most of the fields as arguments in the `openshift-pipelines-controller` deployment under the `openshift-pipelines-controller` container. The {pipelines-shortname} Operator also updates the `buckets` field in the `config-leader-election` configuration map under the `openshift-pipelines` namespace.

If you do not specify the values, the {pipelines-shortname} Operator does not update those fields and applies the default values for the {pipelines-shortname} controller.

[NOTE]
====
If you modify or remove any of the performance fields, the {pipelines-shortname} Operator updates the `openshift-pipelines-controller` deployment and the `config-leader-election` configuration map (if the `buckets` field changed) and re-creates `openshift-pipelines-controller` pods.
====

.Modifiable fields for tuning {pipelines-shortname} performance
[options="header"]
|===

| Name | Description | Default value for the {pipelines-shortname} controller

| `disable-ha` | Enable or disable the high availability (HA) support. By default, the HA support is enabled. | `false`

> | `buckets` | The number of buckets used to partition key space of each reconciler.

Reviewer 3: ```suggestion
| `buckets` | The number of buckets used to partition the key space for each reconciler.
```
- add definite article
- 'for' is more idiomatic here

> Each of the replicas uses these buckets. The instance that owns a bucket handles the reconciling of the keys partitioned into that bucket. The maximum value is `10` | `1`

Reviewer 3: ```suggestion
Each of the replicas uses these buckets. The instance that owns a bucket reconciles the keys partitioned into that bucket. The maximum value is `10` | `1`
```
- fix awkward noun-based phrasing `reconciling of`

> | `threads-per-controller` | The number of threads (workers) to use when the workqueue of the {pipelines-shortname} controller is processed. | `2`

Reviewer 3: Should `workqueue` be `work queue`? If `workqueue` isn't the name of a technical feature, it should be two words.

> | `threads-per-controller` | Specifies the number of threads (workers) to use when the workqueue of the {pipelines-shortname} controller is processed. | `2`

Reviewer 2: Remove "specifies" here and "indicates" in the next one, please

| `kube-api-qps` | The maximum queries per second (QPS) to the cluster master from the REST client. | `5.0`

> | `kube-api-burst` | The maximum burst for throttle. | `10`

Reviewer 3: ```suggestion
| `kube-api-burst` | The maximum burst for a throttle. | `10`
```

= Performance tuning using TektonConfig CR

You can modify the fields under the `.spec.pipeline.performance` parameter in the `TektonConfig` custom resource (CR) to change high availability (HA) support and performance configuration for the {pipelines-shortname} controller.

.Example: `TektonConfig` performance fields
[source,yaml]
----
apiVersion: operator.tekton.dev/v1alpha1
kind: TektonConfig
metadata:
  name: config
spec:
  pipeline:
    performance:
      disable-ha: false
      buckets: 1
      threads-per-controller: 2
      kube-api-qps: 5.0
      kube-api-burst: 10
----

The fields are optional. If you set them, the {pipelines-title} Operator includes most of the fields as arguments in the `openshift-pipelines-controller` deployment under the `openshift-pipelines-controller` container. The {pipelines-shortname} Operator also updates the `buckets` field in the `config-leader-election` configuration map under the `openshift-pipelines` namespace.

If you do not specify the values, the {pipelines-shortname} Operator does not update those fields and applies the default values for the {pipelines-shortname} controller.

[NOTE]
====

> If you modify or remove any of the performance fields, the {pipelines-shortname} Operator updates the `openshift-pipelines-controller` deployment and the `config-leader-election` configuration map (if the `buckets` field changed), and recreates `openshift-pipelines-controller` pods.

Reviewer 2: Suggest removing the comma before "and" and adding a dash "re-creates".

Author: I incorporated the rest of the suggestions.

Reviewer 2: I have sent the IBM Style Guide reference in Slack as the document is not available to the general public.

====

> The following are the modifiable fields:

Reviewer 2: Suggest making this a table title instead (a .something right before the table, no empty lines, turns into a table title) and rewording:

.Configuration fields for tuning {pipelines-shortname} performance

[options="header"]
|===

> | Name | Description | Default value for the pipeline controller

Reviewer 2: {pipelines-shortname} controller

> | `disable-ha` | Enable or disable the HA support. By default, the HA support is enabled. | `false`

Reviewer 2: Suggestion: expand "HA" to "High Availability (HA)" at first mention.

> | `buckets` | Indicates the number of buckets used to partition key space of each reconciler.

Reviewer 2: Remove "Indicates" - just "The number...". ("Indicates" is just outright wrong, this is a configuration field not a monitoring field). I would prefer to remove the passive voice if possible.

> The given number of replicas uses the specified number of buckets. The owner of a bucket handles the reconciling of the keys partitioned into that bucket. The maximum value is `10` | `1`

Reviewer 2: Could we still expand "owner" to "the instance that owns"? Usually "owner" is a user

> The given number of replicas uses these buckets. The instance that owns a bucket handles the reconciling of the keys partitioned into that bucket. The maximum value is `10` | `1`

Reviewer 2: "Each of the replicas" or "All of the replicas, combined" - depending on which is true. "The given number" is still somewhat unclear.

> | `threads-per-controller` | Specifies the number of threads (workers) to use when the workqueue of the pipeline controller is proccessed. | `2`

Reviewer 2: typo - "processed".

| `kube-api-qps` | Indicates the maximum queries per second (QPS) to the cluster master from the REST client. | `5.0`

| `kube-api-burst` | The maximum burst for throttle. | `10`

|===

[NOTE]
====
The {pipelines-shortname} Operator does not control the number of replicas of the pipeline controller. The `replicas` setting of the deployment determines the number of replicas. For example, to change the number of replicas to 3, enter the following command:

[source,terminal]
----
$ oc --namespace openshift-pipelines scale deployment openshift-pipelines-controller --replicas=3
----
====

[IMPORTANT]
====

> The `kube-api-qps` and `kube-api-burst` fields are multiplied by 2 in the pipeline controller. For example, if the `kube-api-qps` and `kube-api-burst` values are `10`, the actual QPS and burst value becomes `20`.

Reviewer 2: "become" not "becomes" as "values" is plural
